File: java_version/dp/maxProductSubArr.java
class Solution {
    int maxProduct(int[] arr) {
        // code here
        int maxProduct=arr[0];
        int minProduct=arr[0];
        int ans=arr[0];
        
        int tempMax=0;
        int tempMin=0;
        
        for(int i=1; i<arr.length; i++)
        {
            int num = arr[i];
            
            tempMax = Math.max(num, Math.max(num*maxProduct, num*minProduct));
            tempMin = Math.min(num, Math.min(num*maxProduct, num*minProduct));
            
            maxProduct=tempMax;
            minProduct=tempMin;
        
        ans=Math.max(ans,maxProduct);
        
        }
        return ans;
    }
}

File: dp/maxProductSubArr.py
class Solution:
    def maxProduct(self, arr):

        maxProd = arr[0]
        minProd = arr[0]
        ans = arr[0]

        for i in range(1, len(arr)):

            num = arr[i]

            tempMax = max(num, num * maxProd, num * minProd)
            tempMin = min(num, num * maxProd, num * minProd)

            maxProd = tempMax
            minProd = tempMin

            ans = max(ans, maxProd)

        return ans
